lib/services/local/presence.go

```diff
@@ -348,6 +349,15 @@ func (s *PresenceService) UpsertNode(ctx context.Context, server types.Server) (
 	if n := server.GetNamespace(); n != apidefaults.Namespace {
 		return nil, trace.BadParameter("cannot place node in namespace %q, custom namespaces are deprecated", n)
 	}
+	if hostname := server.GetHostname(); hostname != "" {
+		// Allow uppercase node hostnames, they aren't valid hostnames
+		// but there's no reason to reject them here. The goal here is
+		// to disallow most special characters.
+		if !utils.IsValidHostname(strings.ToLower(hostname)) {
```

Comment:
Adding new validation requirements here may cause a cache crash loop if there are existing node resources that fail to pass validation already in the backend because `PresenceService.UpsertNode` is reused by the cache.

We should instead perform the check [here](https://github.com/gravitational/teleport/blob/9c108fec44a2649eac8e5cdfb0709bc48e988cb5/lib/inventory/controller.go#L599) and [here](https://github.com/gravitational/teleport/blob/9c108fec44a2649eac8e5cdfb0709bc48e988cb5/lib/auth/auth_with_roles.go#L1016) (heartbeat input point and manual creation input point respectively).

Reply:
Thanks, moved validation to the suggested functions